### PR TITLE
Use `win32-background-launcher` to start the daemon task

### PR DESCRIPTION
**Pull request by anjannath**
Needs:
- [x] https://github.com/crc-org/win32-background-launcher/pull/1
- [x] release `v0.0.0.1` of `win32-background-launcher`

### CODE REVIEW

**Comment by anjannath** [comment]
/hold 

till required tasks are finished as mentioned in https://github.com/crc-org/crc/pull/3901#issue-1970038820 and https://github.com/crc-org/crc/issues/3879#issue-1951516815

**Comment by gbraad** [comment]
can we rename the application on Windows to `crc-background-launcher` instead of the cryptic `win32-...` 
or `background-launcher`. This way people can identify where it belongs to, as paths are not shown in Task Manager.

**Comment by adrianriobo** [comment]
Tested on windows 11 https://crcqe-asia.s3.amazonaws.com/nightly/ocp/4.14.3/20231123/11-pro/qe-results/e2e-non-ux.xml cluster behaves as expected

**Comment by anjannath** [comment]
> can we rename the application on Windows to `crc-background-launcher` instead of the cryptic `win32-...` or `background-launcher`. This way people can identify where it belongs to, as paths are not shown in Task Manager.

yes, can do, first have to first change the name of the executable in crc-org/win32-background-launcher

edit: created https://github.com/crc-org/win32-background-launcher/pull/3

**Comment by cfergeau** [comment]
> can we rename the application on Windows to `crc-background-launcher` instead of the cryptic `win32-...` or `background-launcher`. This way people can identify where it belongs to, as paths are not shown in Task Manager.

Well the launcher is generic and has nothing crc specific. Dunno if there's a way to provide a description to indicate which process it launched?

File: pkg/crc/constants/constants.go [hunk 1]
```diff
@@ -22,18 +22,19 @@ const (
...
-	DaemonVsockPort           = 1024
-	DefaultPodmanNamedPipe    = `\\.\pipe\crc-podman`
-	RootlessPodmanSocket      = "/run/user/1000/podman/podman.sock"
-	RootfulPodmanSocket       = "/run/podman/podman.sock"
+	ConfigFile                    = "crc.json"
+	LogFile                       = "crc.log"
+	DaemonLogFile                 = "crcd.log"
+	CrcLandingPageURL             = "https://console.redhat.com/openshift/create/local" // #nosec G101
+	DefaultAdminHelperURLBase     = "https://github.com/crc-org/admin-helper/releases/download/v%s/%s"
+	CRCWin32BackgroundLauncherURL = "https://github.com/crc-org/win32-background-launcher/releases/download/v%s/%s"
```
Comment (cfergeau) [inline pkg/crc/constants/constants.go]: I don't think the `CRC` prefix is really useful here, and I'd also drop "win32". This does not need to be exported.

File: pkg/crc/constants/constants.go [hunk 2]
```diff
@@ -47,6 +48,8 @@ const (
 
 	OpenShiftIngressHTTPPort  = 80
 	OpenShiftIngressHTTPSPort = 443
+
+	Win32BackgroundLauncherExeName = "win32-background-launcher.exe"
```
Comment (cfergeau) [inline pkg/crc/constants/constants.go]: For admin-helper, we call it `Executable`, not `ExeName`

File: pkg/crc/preflight/preflight_windows.go [hunk 3]
```diff
@@ -100,13 +100,13 @@ var daemonTaskChecks = []Check{
 		labels: labels{Os: Windows},
 	},
 	{
-		configKeySuffix:    "check-daemon-task-posh-script-present",
-		checkDescription:   "Checking if the daemon task powershell script is present",
-		check:              checkDaemonPoshScript,
```
Comment (cfergeau) [inline pkg/crc/preflight/preflight_windows.go:105]: This was removed in the previous commit, would be nice if each commit builds.

**Review by cfergeau — COMMENTED** [review]
overall looks fine, a few comments here and there.

File: pkg/crc/constants/constants.go [hunk 4]
```diff
@@ -187,6 +194,11 @@ func GetKubeAdminPasswordPath() string {
 	return filepath.Join(MachineInstanceDir, DefaultName, "kubeadmin-password")
 }
 
+func GetCRCWindowsBackgroundLauncherDownloadURL() string {
```
Comment (cfergeau) [inline pkg/crc/constants/constants.go]: GetWin32BackgroundLauncher...?

File: pkg/crc/cache/cache.go [hunk 5]
```diff
@@ -20,6 +20,7 @@ type Cache struct {
 	executablePath string
 	archiveURL     string
 	version        string
+	nameMismatch   bool
```
Comment (cfergeau) [inline pkg/crc/cache/cache.go]: `ignoreNameMismatch` would be more explicit.

File: pkg/crc/cache/cache.go [hunk 6]
```diff
@@ -33,8 +34,8 @@ func (e *VersionMismatchError) Error() string {
 	return fmt.Sprintf("%s version mismatch: %s expected but %s found in the cache", e.ExecutableName, e.ExpectedVersion, e.CurrentVersion)
 }
 
-func newCache(executablePath string, archiveURL string, version string, getVersion func(string) (string, error)) *Cache {
-	return &Cache{executablePath: executablePath, archiveURL: archiveURL, version: version, getVersion: getVersion}
+func newCache(executablePath string, archiveURL string, version string, nameMismatch bool, getVersion func(string) (string, error)) *Cache {
```
Comment (cfergeau) [inline pkg/crc/cache/cache.go]: Not sure I'd add this arg  by default, might be better to explicitly set it for the background launcher after creating the cache with the default values (`cache.IgnoreNameMismatch()` or `cache.ignoreNameMismatch = true`)

File: pkg/crc/preflight/preflight_daemon_task_check_windows.go [hunk 7]
```diff
@@ -120,7 +86,16 @@ func fixDaemonTaskInstalled() error {
...
+	crcBinPath, err := os.Executable()
+	if err != nil {
+		return err
+	}
+
+	if !crcos.FileExists(constants.Win32BackgroundLauncherPath()) {
+		return fmt.Errorf("Missing background launcher binary at: %s", constants.Win32BackgroundLauncherPath())
+	}
+
+	binPathWithArgs := fmt.Sprintf(`"%s" daemon`, crcBinPath)
```
Comment (cfergeau) [inline pkg/crc/preflight/preflight_daemon_task_check_windows.go:99]: This used to be `daemonCmdArgs := `daemon --log-level debug`, is it intentional that you removed the debug logs?

File: pkg/crc/cache/cache_windows.go [hunk 8]
```diff
@@ -23,4 +23,6 @@ func NewWin32BackgroundLauncherCache() *Cache {
 			return strings.TrimSpace(stdOut), nil
 		},
 	)
+	cache.ignoreNameMismatch = true
```
Comment (cfergeau) [inline pkg/crc/cache/cache_windows.go:28]: Might be worth an explanatory comment? `upstream binary is win32-..., we want to name it crc-...`?

**Review by cfergeau — APPROVED** [review]
2 minor comments, but fine with me with or without changes
/lgtm
/approve

File: pkg/crc/preflight/preflight_daemon_task_check_windows.go (hunk 7 above)
Comment (anjannath) [reply pkg/crc/preflight/preflight_daemon_task_check_windows.go:99]: yes, we have no way to look at the stdout of the process, and logs to the log file `crcd.log` is always at debug/trace level, so this flag was not useful

**Pull request merged**